src/sentry/sentry_metrics/consumers/indexer/batch.py

```diff
@@ -448,123 +449,125 @@ def reconstruct_messages(
-                routing_payload = RoutingPayload(
-                    routing_header={"org_id": org_id},
-                    routing_message=kafka_payload,
+            with metrics.timer("metrics_consumer.reconstruct_messages.build_new_payload"):
+                if self.__should_index_tag_values:
+                    # Metrics don't support gauges (which use dicts), so assert value type
+                    value = old_payload_value["value"]
+                    assert isinstance(value, (int, float, list))
+                    new_payload_v1: Metric = {
+                        "tags": new_tags,
+                        # XXX: relay actually sends this value unconditionally
+                        "retention_days": old_payload_value.get("retention_days", 90),
+                        "mapping_meta": output_message_meta,
+                        "use_case_id": old_payload_value["use_case_id"].value,
+                        "metric_id": numeric_metric_id,
+                        "org_id": old_payload_value["org_id"],
+                        "timestamp": old_payload_value["timestamp"],
+                        "project_id": old_payload_value["project_id"],
+                        "type": old_payload_value["type"],
+                        "value": value,
+                        "sentry_received_timestamp": sentry_received_timestamp,
+                    }
+
+                    new_payload_value = new_payload_v1
+                else:
+                    # When sending tag values as strings, set the version on the payload
+                    # to 2. This is used by the consumer to determine how to decode the
+                    # tag values.
+                    new_payload_v2: GenericMetric = {
+                        "tags": cast(Dict[str, str], new_tags),
+                        "version": 2,
+                        "retention_days": old_payload_value.get("retention_days", 90),
+                        "mapping_meta": output_message_meta,
+                        "use_case_id": old_payload_value["use_case_id"].value,
+                        "metric_id": numeric_metric_id,
+                        "org_id": old_payload_value["org_id"],
+                        "timestamp": old_payload_value["timestamp"],
+                        "project_id": old_payload_value["project_id"],
+                        "type": old_payload_value["type"],
+                        "value": old_payload_value["value"],
+                        "sentry_received_timestamp": sentry_received_timestamp,
+                    }
+                    if aggregation_option := get_aggregation_option(old_payload_value["name"]):
+                        new_payload_v2["aggregation_option"] = aggregation_option.value
+
+                    new_payload_value = new_payload_v2
+
+                kafka_payload = KafkaPayload(
+                    key=message.payload.key,
+                    value=rapidjson.dumps(new_payload_value).encode(),
```

Comment:
I sort of wonder if the time `rapidjson.dumps` takes should be measured independently. It's the only thing I see in the `metrics_consumer.reconstruct_messages.build_new_payload` path that I would expect to take any time

Reply:
That's fair. I can break down the `build_new_payload` metric down further